integrated_channels/canvas/client.py

```diff
@@ -378,37 +378,65 @@ def _handle_canvas_assignment_retrieval(
-            raise ClientError(
-                "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
-                    resp.text,
-                ),
-                resp.status_code
-            )
+        more_pages_present = True
+        current_page_count = 0
+        assignment_id = ''
 
-        assignments_resp = resp.json()
-        assignment_id = None
-        for assignment in assignments_resp:
-            try:
-                if assignment['integration_id'] == integration_id:
-                    assignment_id = assignment['id']
-                    break
-            # The validation check above should ensure that we have a 200 response from Canvas, but sanity catch if we
-            # have a unexpected response format
-            except (KeyError, ValueError, TypeError) as error:
+        # current_page_count serves as a timeout, limiting to a max of 150 pages of requests
+        while more_pages_present and current_page_count < 150:
+            if resp.status_code >= 400:
                 raise ClientError(
                     "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
                         resp.text,
                     ),
                     resp.status_code
-                ) from error
+                )
+
+            assignments_resp = resp.json()
+            for assignment in assignments_resp:
+                try:
+                    if assignment['integration_id'] == integration_id:
+                        assignment_id = assignment['id']
+                        break
+
+                # The integration ID check above should ensure that we have a 200 response from Canvas,
+                # but sanity catch if we have a unexpected response format
+                except (KeyError, ValueError, TypeError) as error:
+                    raise ClientError(
+                        "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
+                            resp.text,
+                        ),
+                        resp.status_code
+                    ) from error
+
+            if not assignment_id:
+                # Canvas pagination headers come back as a string ie:
+                # 'headers': {
+                #     'Link': '<{assignment_url}?page=2&per_page=10>; rel="current",' \
+                #             '<{assignment_url}?page=1&per_page=10>; rel="prev",' \
+                #             '<{assignment_url}?page=1&per_page=10>; rel="first",' \
+                #             '<{assignment_url}?page=2&per_page=10>; rel="last"' \
+                # }
+                # so we have to parse out the linked list of assignment pages
+                assignment_page_results = resp.headers['Link'].split(',')
+                current_page = assignment_page_results[0].split(';')[0].strip('<>')
+                last_page = assignment_page_results[3].split(';')[0].strip('<>')
```

Comment:
I see this in their doc `rel="last" may also be excluded if the total count is too expensive to compute on each request.`. Guess we won't hit this often. But is there a chance the 'last' page will be absent in some cases?

Reply:
this is an interesting point. Let me investigate behaviors but yeah I should probably account for this.